### PR TITLE
feat: extended bounty

### PR DESCRIPTION
**Pull request by nksazonov**

### CODE REVIEW

File: contracts/duckies/Garden.sol [hunk 1]
```diff
@@ -121,52 +119,6 @@ contract Garden is Initializable, AccessControlUpgradeable, UUPSUpgradeable {
...
-
-		for (uint i = 0; i < referralPayouts.length; i++) {
-			referralPayouts[i] /= halvingDivider;
-		}
-		return referralPayouts;
-	}
-
-	// -------- Halving --------
-
-	function halving() public view returns (uint8) {
```
Comment (mod) [inline contracts/duckies/Garden.sol:153]: I suggest we remove halving and clean the code so we can merge, before adding changes

File: contracts/duckies/RandomUpgradeable.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+contract RandomUpgradeable is Initializable {
```
Comment (mod) [inline contracts/duckies/RandomUpgradeable.sol:6]: I don't understand what is this file for

File: contracts/duckies/Gene.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+library Gene {
+    uint8 public constant BYTES_PER_TRAIT = 8;
+
+    enum Classes {
+        Common,
+        Rare,
+        Epic,
+        Legendary,
+        SuperLegendary,
+        Zombie
```
Comment (mod) [inline contracts/duckies/Gene.sol:26]: Zombeak,
also it's not a class

File: contracts/duckies/Gene.sol [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        Element,
+        Eyes,
+        Beak,
+        Wings,
+        Firstname,
+        Lastname,
+        Temper,
+        Peculiarity
+    }
+
```
Comment (mod) [inline contracts/duckies/Gene.sol:44]: how can we cover with tests?

File: contracts/duckies/Garden.sol [hunk 5]
```diff
@@ -23,33 +25,32 @@ contract Garden is Initializable, AccessControlUpgradeable, UUPSUpgradeable {
...
+		TransferBountyToken, // encodedData: abi.encode(address, uint256, uint8[5]) - encoded token address, amount to pay, referrer payouts
+		MintDucklingsNFT // encodedData: 0x (empty). May be extended in the future.
+	}
+
 	// Bounty Message for signature verification
 	struct Bounty {
-		uint256 amount;
-		address tokenAddress;
+		BountyType type_;
+		bytes encodedData; // specific to BountyType
```
Comment (mod) [inline contracts/duckies/Garden.sol:38]: I like the flexibility, but, could we read data into a structure, for example we call it
```
struct PayoutParams {
    address token;
    uint256  amount;
    uint8      payouts;
}

struct FactoryParams {
    address token;        // NFT contract address
    uint256  quantity;    // Size of card pack
    uint8      collection;  // index of collection
}

struct MeldParams {
    bytes32[5]  family;  // Array of DNA of the cards, or TokenID, need to verify ownership
}

// Then in the switch case we unpack into
// memory MeldParams args = unpack(data);
// meldCards(args);
```

**Review by mod — CHANGES_REQUESTED** [review]
Let's proceed step by step,
We can have an empty factory now, let's focus on the Garden code for now
I want to improve readability of the code have having explicit methods and Typed params
See my comment

File: contracts/duckies/Garden.sol (hunk 5 above)
Comment (nksazonov) [reply contracts/duckies/Garden.sol:38]: Sure, of course we can.

File: contracts/duckies/Gene.sol (hunk 4 above)
Comment (nksazonov) [reply contracts/duckies/Gene.sol:44]: That generated values should be in the range [0, Max value].

File: contracts/duckies/Garden.sol (hunk 1 above)
Comment (nksazonov) [reply contracts/duckies/Garden.sol:153]: Yes, I agree. Then, I'll separate this PR into 2.

File: contracts/duckies/RandomUpgradeable.sol (hunk 2 above)
Comment (nksazonov) [reply contracts/duckies/RandomUpgradeable.sol:6]: This is a peculiarity of Upgradeable contracts. As Random also holds storage variables and Garden inherits from it, Random should be declared Upgradeable and inherit from Initializable.

**Pull request closed without merging**